schemachanger: Fix ALTER TABLE ADD COLUMN j INT, DROP COLUMN j

Xiang-Gu opened the pull request:
Previously, the declarative schema changer errors for ALTER TABLE stmt where we add and then drop the same column. This commit fixes that.

Epic: None
Fixes #110589
Release note: None

Xiang-Gu committed:
schemachanger: Fix ALTER TABLE ADD COLUMN j INT, DROP COLUMN j

Previously, the declarative schema changer errors for ALTER TABLE stmt
where we add and then drop the same column. The reason is because the
mechanism we use to handle ADD COLUMN, DROP COLUMN, ALTER PK is flawed.
Namely, the mechanism is a two-step process:

1. We inflate a chain of primary indexes denoted by
`old` <-- `inter1` <-- `inter2` <-- `final`
  - ADD COLUMN: add `col` to `inter1`, `inter2`, and `final`
  - DROP COLUMN: drop `col` from `final`
  - ALTER PK: alter primary column(s) in `inter2` and `final`

2. At the end of each statement, we remove redundant primary indexes
because not all `inter1`, `inter2`, and `final` are needed.

Unfortunately, this mechanism won't work for statement like
`ADD COLUMN j INT DEFAULT 30, DROP COLUMN j` because
 - `ADD COLUMN j` adds `j` to `inter1`, `inter2`, and `final`, but
 - `DROP COLUMN j` only removes `j` from `final`

The fix is therefore to amend how we handle `DROP COLUMN` so that, if
the column to drop is newly added in the same transaction, which means
it is added to `inter1`, `inter2`, `final` (and possibly `old`), we will
detect such cases and drop the column from those indexes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110851)
<!-- Reviewable:end -->

Xiang-Gu commented on pkg/sql/alter_table.go:
```diff
@@ -1831,7 +1831,7 @@ func dropColumnImpl(
 	}
 	if !found {
-		return nil, pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
+		return nil, pgerror.Newf(pgcode.FeatureNotSupported,
```
This is to please `TestCompareLegacyAndDeclarative` in which we assert that if some DDL returns an error in legacy schema changer but succeeds in declarative schema changer, the error must be FeatureNotSupported error.

fqazi reviewed (changes requested):
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 500 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-Nehm52BENMjtjnE-j2L:-Nehm52CDew4fzPvQURn:bucgepf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L500)):*
> ```Go
> 	) {
> 		if e.IndexID == indexID && e.ColumnID == columnID {
> 			current, target, indexColumn = c, t, e
> ```

Should we be paranoid and assert there is only one?
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 3221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NehmNF88pxMQZ-ShG3Z:-NehmNF88pxMQZ-ShG3_:b-wqfa8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/logictest/testdata/logic_test/alter_table#L3221)):*
> ```
> skipif config local-mixed-22.2-23.1
> statement ok
> ALTER TABLE t_99035 ADD COLUMN j INT, DROP COLUMN j;  -- noop
> ```

Let's get an end-to-end test for this.


<!-- Sent from Reviewable.io -->

chrisseto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dikshant and @Xiang-Gu)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-Nehmfc29AthHz1j9sYo):*
Could you add a summary of what the issue was and how it's being fixed within this commit?

Doing so it's helpful as it let's reviewers compare your description of the bug and fix to the code changes rather than having to reverse engineer the issue from this change.
___
*[`pkg/sql/alter_table.go` line 1834 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110851#gh-1329254866:-Nehiegz7blfmER7-ZoO:b-7u0pdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cb41e9fbe6c57053a815dbeb141dfe62d0c7287/pkg/sql/alter_table.go#L1834)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

This is to please `TestCompareLegacyAndDeclarative` in which we assert that if some DDL returns an error in legacy schema changer but succeeds in declarative schema changer, the error must be FeatureNotSupported error.
</blockquote></details>

Do we have any guarantees about backwards compatibility of error codes? I can see an argument either way in terms of what code this one should be.

@dikshant , WDYT?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 488 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NehlH7dDNObiNy5pdiE:-NehlH7dDNObiNy5pdiF:bksro4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L488)):*
> ```Go
> 	dropStoredColumnFromPrimaryIndex(b, tbl.TableID, inflatedChain.finalSpec.primary, col)
> 
> 	// If `col` already exists in `old`, then we can return now (after dropping
> ```

This comment feels pretty jargon-y to me and doesn't mesh super well with the flow of control.
After about 3-4 reads and poking around the code, I _think_ I understand what it's saying. 

How would you feel about restructuring this to something like:
```go
// if `col` exists and is public in `old`, drop it from the primary index as normal.
if e != nil && current == scpb.Status_PUBLIC {
    dropStoredColunFromPrimaryIndex(...)
    return
}

// If `col` isn't present or public in `old`, it's been added and dropped within the same transaction.
// In such a case, we need to ensure it's not present in .... because ....
for _, _ := range []{} {
   
}
```

In doing so, we could remove the `fromID == 0` check from `dropStoredColumnFromPrimaryIndex`. That bit also reads a bit strange to me but that might be the convention in this package.

Could you explain a bit more about why we need to drop the column from all these other states in this specific case?


<!-- Sent from Reviewable.io -->

chrisseto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dikshant, @fqazi, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 500 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-Nehm52BENMjtjnE-j2L:-NehnABwD4iXjgqIIUQ9:b-k4jb93) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L500)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Should we be paranoid and assert there is only one?
</blockquote></details>

Forgot to ask this same question! +1 to Faizan's question.


<!-- Sent from Reviewable.io -->

chrisseto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dikshant, @fqazi, and @Xiang-Gu)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 3221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NehmNF88pxMQZ-ShG3Z:-NehnLIY2g37kpcrcCBL:b-y6jqfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/logictest/testdata/logic_test/alter_table#L3221)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Let's get an end-to-end test for this.
</blockquote></details>

Can we also add a test for performing the add and drop within the same transaction but across multiple statements?

```sql
ALTER TABLE t ADD COLUMN j int;
-- some unrelated DDL
ALTER TABLE t DROP COLUMN j;
```


<!-- Sent from Reviewable.io -->

Xiang-Gu committed:
schemachanger: Add an end-to-end test

This commit augments an existing end-to-end test case of alter table
with multiple commands with two more commands `ADD COLUMN p INT, DROP p`.
We're glad to see that the resulting plan remains the same. This change
is separated from the previous commit mostly to make the previous commit
(the main commit) less crowded.

Release: None

Xiang-Gu reviewed (commented):
Hi @fqazi and @chrisseto, I spent today looking into this more. Your feedback is very valuable, and while I think I can address most of them, there is one case that is surprisingly challenging:
```
BEGIN;
ALTER TABLE t ADD COLUMN j INT;
ALTER TABLE t DROP COLUMN j;
COMMIT
```

To illustrate the difficulty, I want to first point out one observation: the first ADD COLUMN in the above example will result in a StatementPhase like the following
```
   ├── StatementPhase
   │    └── Stage 1 of 1 in StatementPhase
   │         ├── 8 elements transitioning toward PUBLIC
   │         │    ├── ABSENT → DELETE_ONLY   Column:{DescID: 104 (t), ColumnID: 2 (j+)}
   │         │    ├── ABSENT → PUBLIC        ColumnName:{DescID: 104 (t), Name: "j", ColumnID: 2 (j+)}
   │         │    ├── ABSENT → PUBLIC        ColumnType:{DescID: 104 (t), ColumnFamilyID: 0 (primary), ColumnID: 2 (j+)}
   │         │    ├── ABSENT → PUBLIC        ColumnDefaultExpression:{DescID: 104 (t), ColumnID: 2 (j+)}
   │         │    ├── ABSENT → BACKFILL_ONLY PrimaryIndex:{DescID: 104 (t), IndexID: 2 (t_pkey+), ConstraintID: 2, TemporaryIndexID: 3, SourceIndexID: 1 (t_pkey-)}
   │         │    ├── ABSENT → PUBLIC        IndexColumn:{DescID: 104 (t), ColumnID: 1 (i), IndexID: 2 (t_pkey+)}
   │         │    ├── ABSENT → PUBLIC        IndexData:{DescID: 104 (t), IndexID: 2 (t_pkey+)}
   │         │    └── ABSENT → PUBLIC        IndexColumn:{DescID: 104 (t), ColumnID: 2 (j+), IndexID: 2 (t_pkey+)}
```
where the new primary index is *eagerly* created and transitioned to BACKFILL_ONLY status. 
Another observation is that today's code that handles ADD COLUMN(s), DROP COLUMN(s), ALTER PK is an "online algorithm/mechanism" in that we exercise the algorithm (which can be summarized as "inflate chain -> modify chain -> deflate chain") fully for each statement in the builder (and therefore they effect in StatementPhase). Such an algorithm becomes very difficult to maintain (bc of its ins and outs) across multiple ADD COLUMN(s), DROP COLUMN(s), ALTER PK statements in one transaction (we didn't foresee this difficulty during initial development bc back then, we (or rather I) most have the canonical example of *one* statement with multiple ADD COLUMN(s), DROP COLUMN(s), ALTER PK in mind).

All these thinking prompted a solution that maybe we should not eagerly exercise the full machinery during StatementPhase, and it will be much much simpler to delay it to PreCommitPhase, where we would know all the columns to add, to drop, and all the PK changes, and we just need to handle creating new primary indexes once (and for all). It will be a simpler implementation and it's easier to reason and prove its correctness. The gist here is to simplify an "online algorithm", which is hard to consider all cases and hard to implement correctly, to an "off algorithm", since there is no need for it to be "online".

For now, I have one question for you two: What do you think of me still polishing up this PR and merging it (so at least we can handle cases like `ALTER TABLE ADD COLUMN, DROP COLUMN` where the add and drop are in one stmt), while acknowledging that there are still cases we cannot handle correctly and fixing it requires much more effort? 
Or, do you guys think we should not merge at all and wait for the big future effort to fix it altogether, if we ever get the chance to?
I personally prefer the former (because, regardless of the gap, it's a strict improvement of today's code). If you think so too, then I've addressed all of your comments and this PR is ready for another look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-:-Neja9-7ExyLX05dnyv1:b3lppid)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @dikshant, and @fqazi)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-Nehmfc29AthHz1j9sYo):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Could you add a summary of what the issue was and how it's being fixed within this commit?

Doing so it's helpful as it let's reviewers compare your description of the bug and fix to the code changes rather than having to reverse engineer the issue from this change.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 500 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-Nehm52BENMjtjnE-j2L:-NeiRv7gA9GXu-6e4fR2:b-ky38rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L500)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Forgot to ask this same question! +1 to Faizan's question.
</blockquote></details>

good point. Done!
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 488 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NehlH7dDNObiNy5pdiE:-NeiPb1M2PLR3vHuWk9r:b-r9wdpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L488)):*
Thanks for the suggestion. It is a good one! 

> In doing so, we could remove the `fromID == 0` check from `dropStoredColumnFromPrimaryIndex`

I think this is still needed because `old` primary index does not have a temporary index associated with it.
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 3221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NehmNF88pxMQZ-ShG3Z:-Nej_zAU52R8WOLedUL6:b-onceiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/logictest/testdata/logic_test/alter_table#L3221)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Can we also add a test for performing the add and drop within the same transaction but across multiple statements?

```sql
ALTER TABLE t ADD COLUMN j int;
-- some unrelated DDL
ALTER TABLE t DROP COLUMN j;
```
</blockquote></details>

@fqazi Augmented an existing end-to-end test case to include `ADD COLUMN, DROP COLUMN`

@chrisseto see my long comment as to why we cannot do that right now


<!-- Sent from Reviewable.io -->

fqazi commented:
> For now, I have one question for you two: What do you think of me still polishing up this PR and merging it (so at least we can handle cases like ALTER TABLE ADD COLUMN, DROP COLUMN where the add and drop are in one stmt), while acknowledging that there are still cases we cannot handle correctly and fixing it requires much more effort?
Or, do you guys think we should not merge at all and wait for the big future effort to fix it altogether, if we ever get the chance to?

@Xiang-Gu It is fair to deal with the transactional case in a separate PR, we aren't planning on supporting that any time soon

chrisseto reviewed (approved):
> For now, I have one question for you two: What do you think of me still polishing up this PR and merging it (so at least we can handle cases like `ALTER TABLE ADD COLUMN, DROP COLUMN` where the add and drop are in one stmt), while acknowledging that there are still cases we cannot handle correctly and fixing it requires much more effort?

Fascinating! I had thought the two cases where the same. Yeah, we can punt the separate statement case into a different PR. Would you mind filing an issue for it?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-:-Nenc3oJ8xJqb4KWlObd:b-2407ao)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dikshant, @fqazi, and @Xiang-Gu)

___
*[`-- commits` line 10 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NencM6o4QYxzJFM3A0R):*
Unrelated to this PR. Is inter1 and inter2 a reference to what you mentioned about being able to perform all operations with at most 2 index rewrites?
___
*[`pkg/sql/explain_ddl.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-Nenco4Z9nFhI5rxvCup:-Nenco4Z9nFhI5rxvCuq:bjalwvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e3178b2f659f62e35928e4ec35f7a59c6bf329/pkg/sql/explain_ddl.go#L91)):*
> ```Go
> 		ExecutionPhase:             scop.StatementPhase,
> 		SchemaChangerJobIDSupplier: func() jobspb.JobID { return 1 },
> 		SkipPlannerSanityChecks:    true,
> ```

Is this related to this PR/commit/issue?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 488 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NehlH7dDNObiNy5pdiE:-NencfU2EY4cxJ-MO8Lp:b-aoz9vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L488)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Thanks for the suggestion. It is a good one! 

> In doing so, we could remove the `fromID == 0` check from `dropStoredColumnFromPrimaryIndex`

I think this is still needed because `old` primary index does not have a temporary index associated with it.
</blockquote></details>

Got it! Can you add a note to that check in `dropIndexColumnFromInternal` in that case?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> Reviewed 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @dikshant, and @Xiang-Gu)

___
*[`pkg/sql/explain_ddl.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-Nenco4Z9nFhI5rxvCup:-NengxcX5Ok6nNMTDUmS:bdk7ku9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e3178b2f659f62e35928e4ec35f7a59c6bf329/pkg/sql/explain_ddl.go#L91)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Is this related to this PR/commit/issue?
</blockquote></details>

Why is this needed?


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
@chrisseto filed https://github.com/cockroachdb/cockroach/issues/110995 to track this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-:-Neo7RXp17gy5tfOKe-h:b9rekp9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @dikshant, and @fqazi)

___
*[`-- commits` line 10 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NencM6o4QYxzJFM3A0R):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Unrelated to this PR. Is inter1 and inter2 a reference to what you mentioned about being able to perform all operations with at most 2 index rewrites?
</blockquote></details>

yes! but it's at most 3 indexes rewrites (that's what we call `inter1`, `inter2`, and `final` throughout the codebase)
___
*[`pkg/sql/explain_ddl.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-Nenco4Z9nFhI5rxvCup:-Neo3YjUEg0KILXDVBz2:b-s06yc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e3178b2f659f62e35928e4ec35f7a59c6bf329/pkg/sql/explain_ddl.go#L91)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Why is this needed?
</blockquote></details>

removed.

Long story: I ran into some internal error (from plannerSanityChecks) when trying to use `EXPLAIN(DDL)`. This is a quick hack to get around it, but it's not related to this PR. 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 488 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-NehlH7dDNObiNy5pdiE:-Neo4elaAZ9pFH3o8srW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d743e02df2853f6650afe398f1024676be098209/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L488)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Got it! Can you add a note to that check in `dropIndexColumnFromInternal` in that case?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great work!

> Reviewed 37 of 37 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110851#-:-Neo8IHPDfMnEEpH6ZP5:b-sap6he)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chrisseto and @dikshant)



<!-- Sent from Reviewable.io -->

dikshant commented:
> Do we have any guarantees about backwards compatibility of error codes? I can see an argument either way in terms of what code this one should be.

It's a soft guarantee as in customers probably don't expect the error code to change under their feet. What is the old vs the new error code?

Xiang-Gu commented:
> It's a soft guarantee as in customers probably don't expect the error code to change under their feet. What is the old vs the new error code? 

@dikshant Old code is `ObjectNotInPrerequisiteState (55000)`, the new proposed code is `FeatureNotSupported (0A000)`.

By the way, for the same stmt, PG code is `UndefinedColumn (42703)`.

dikshant commented:
@Xiang-Gu what is keeping us from making that error code the same as Postgres?

Xiang-Gu commented:
@dikshant and I discussed the error code change offline; we agreed that this is okay since they already don't match today.

bors r+
